### PR TITLE
drivers/usbdev_synopsys_dwc2: add EFM32 support

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides the changes for the Synopsys USB OTG IP core DWC2 driver for EFM32 MCUs. It also provides the changes of the board definition for `stk3600` and `stk3700` for testing.

### Testing procedure

`tests/usbus_hid` should work on the EFM32 boards `stk3600` (EFM32LG family) and `stk3700` (EFM32GG family).

It is already tested for a `sltb009a` board (EFM32GG12 family).

### CODE REVIEW

File: cpu/efm32/Makefile.dep [hunk 1]
```diff
@@ -13,4 +13,9 @@ endif
 # include CPU family module
 USEMODULE += cpu_$(CPU_FAM)
 
+ifneq (,$(filter periph_usbdev,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_usbdev
```
Comment (jue89) [inline cpu/efm32/Makefile.dep:17]:
```suggestion
```

I'd say features are already evaluated at this point? It's too late to require further features.

I'd say a `FEATURE_REQUIRED += periph_usbdev` in some Makefile already resulted into `periph_usbdev` being part of `USEMODULE`.

**Comment by chrysn** [comment]
Not much time for reviewing, but testing I can fit in, on stk3700 and a custom board.

File: cpu/efm32/Makefile.dep (hunk 1 above)
Comment (gschorcht) [reply cpu/efm32/Makefile.dep:17]: You are right.

**Comment by gschorcht** [comment]
I still have some work to do. I had to realize that the EFM32GG family does not have the USHFRCO clock that is used as USB_CLK on EFM32GG12B. That means we have to use different clock sources for different families :worried:

File: cpu/esp32/Kconfig.esp32s2 [hunk 2]
```diff
@@ -26,7 +26,7 @@ config CPU_FAM_ESP32S2
     select MODULE_PS if MODULE_SHELL
     select MODULE_PTHREAD if MODULE_CPP
     select MODULE_RTT_RTC if HAS_PERIPH_RTT && MODULE_PERIPH_RTC
-    select MODULE_USBDEV_SYNOPSYS_DWC2 if HAS_PERIPH_USBDEV && MODULE_PERIPH_USBDEV
```
Comment (dylad) [inline cpu/esp32/Kconfig.esp32s2:29]: It would be better to split ESP32 to another PR.
Changes are rather small so this should be quick.

**Comment by gschorcht** [comment]
I had to add separate USB clock handling for EFM32GG and EFM32LG families. Unfortunatly, I can't test it because I don't have the hardware.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c [hunk 3]
```diff
@@ -578,15 +580,20 @@ static void _sleep_periph(const dwc2_usb_otg_fshs_config_t *conf)
...
 #endif
 }
 
 static void _wake_periph(const dwc2_usb_otg_fshs_config_t *conf)
 {
-#ifdef STM32_USB_OTG_CID_1x
+#if defined(STM32_USB_OTG_CID_1x)
     pm_block(STM32_PM_STOP);
+#elif defined(MCU_EFM32)
+    pm_block(1);
```
Comment (jue89) [inline drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:596]:
```suggestion
    pm_block(PM_MODE_EM2);
```

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c [hunk 4]
```diff
@@ -578,15 +580,20 @@ static void _sleep_periph(const dwc2_usb_otg_fshs_config_t *conf)
     /* Unblocking STM32_PM_STOP during suspend on the stm32f446 breaks
      * while (un)blocking works on the stm32f401, needs more
      * investigation with a larger set of chips */
-#ifdef STM32_USB_OTG_CID_1x
+#if defined(STM32_USB_OTG_CID_1x)
     pm_unblock(STM32_PM_STOP);
+#elif defined(MCU_EFM32)
+    pm_block(0);
+    pm_unblock(1);
```
Comment (jue89) [inline drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:587]: I'd suggest to name those power modes. I added a commit to my PR. Feel free to cherry-pick:
https://github.com/RIOT-OS/RIOT/pull/18760/commits/6ec0135166544013116a139a4af99773c6f10e6f

```suggestion
    pm_block(PM_MODE_EM3);
    pm_unblock(PM_MODE_EM2);
```

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c [hunk 5]
```diff
@@ -674,6 +684,40 @@ static void _usbdev_init(usbdev_t *dev)
     if (usb_new_phy(&phy_conf, &phy_hdl) != ESP_OK) {
         LOG_ERROR("usbdev: Install USB PHY failed\n");
     }
+
+#elif defined(MCU_EFM32)
+
+    /* Block EM2 and EM3. In EM2, most USB core registers are reset and the
+     * FIFO content is lost. EM2 is unblocked during USB suspend */
+    pm_block(1);
```
Comment (jue89) [inline drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:692]:
```suggestion
    pm_block(PM_MODE_EM2);
```

**Review by jue89 — CHANGES_REQUESTED** [review]
Some comments regarding `pm_layered`. Haven't tested it.

Is there maybe a `pm_unblock(PM_MODE_EM3)` missing? I can just find a `pm_block(PM_MODE_EM3)`

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 3 above)
Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:596]: This will only work when your PR #18760 is merged before. Alternatively, I could cherry-pick your commit https://github.com/RIOT-OS/RIOT/pull/18760/commits/6ec0135166544013116a139a4af99773c6f10e6f

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 4 above)
Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:587]: Ah ok, this comment answers my questions. I should have read all comments before I answer :sunglasses:

**Comment by gschorcht** [comment]
> Is there maybe a pm_unblock(PM_MODE_EM3) missing? I can just find a pm_block(PM_MODE_EM3)

Hm, I'm not familar with EFM32 and its power managed. If I'm understand it right, USB will not work at all in EM3 and is not wakable from this power mode. That's why I thought that we have to block it if `periph_usbdev` is used.

**Comment by jue89** [comment]
> > Is there maybe a pm_unblock(PM_MODE_EM3) missing? I can just find a pm_block(PM_MODE_EM3)
> 
> Hm, I'm not familiar with EFM32 and its power managed. If I'm understand it right, USB will not work at all in EM3 and is not wakable from this power mode. That's why I thought that we have to block it if `periph_usbdev` is used.

`pm_unblock()` and `pm_block()` are reference counting blockers in a given pm mode.

Can you outline the states the drivers might be in? How are `_wake_periph()`, `_sleep_periph()` and `_usbdev_init()` used? Which power modes are required in which state?

**Comment by gschorcht** [comment]
> Can you outline the states the drivers might be in? How are _wake_periph(), _sleep_periph() and _usbdev_init() used?

In `_usbdev_init` the USB IP core is initialized and the USB port is activated. Because the USB core uses either HFCLK or USHFRCO as clock source, only EM0 and EM1 are allowed while the USB port is active. For this reason `PM_MODE_EM2` is blocked. When the USB core detects a suspend on the USB bus, a suspend interrupt is triggered and `_sleep_periph` is called. After switching the USB core clock source to LFXO or LFRCO (not implemented yet :worried:), EM2 can be entered. Therefore `PM_MODE_EM2` is enabled in `_sleep_periph`. In this power mode, the USB core is able to detect a resume signal or a remote wakeup on the USB bus and triggers a wakeup interrupt to call `_wake_periph`. In `_wakeup_periph` the USB core clock source should be switched back to HFCLK or USHFRCO (not yet implemented) and `PM_MODE_EM2` must be disabled.

Because USB is not working in EM3, this power mode must not be entered and should be blocked forever once the USB device is initialized. So it should be done in `_usbdev_init` and not in `_sleep_periph` as done for STM32.

**Comment by jue89** [comment]
I'd suggest just to call `pm_block(PM_MODE_PM2)` once in `_usbdev_init()` until the clock source switching is implemented.

Once the clock source switching is implemented, we need the following logic:

- `_usbdev_init()` blocks `PM_MODE_PM2` -> PM1 is the lowest allowed pm mode
- `_sleep_periph()` unblocks `PM_MODE_PM2` and blocks `PM_MODE_PM3` -> PM2 ist the lowest allowed pm mode
- `_wake_periph()` unblocks `PM_MODE_PM3` and blocks `PM_MODE_PM2` -> PM1 ist the lowest allowed pm mode

**Comment by gschorcht** [comment]
While trying to fix the remaining compilation issue of `bootloaders/riotboot_dfu` for EFM32 usbdev by increasing the `RIOTBOOT_LEN`, I realized that it is working for the `nucleo-f767zi` board but not for the `stm32f429i-disc1` board. On `stm32f429i-disc1` board, it crashes here https://github.com/RIOT-OS/RIOT/blob/b4b8f2bcd399ba522bc4284e4e55f3e5a7c7a0cd/sys/riotboot/usb_dfu.c#L38 when accessing `*reset_addr` due to a hard fault. Has someone observed something like that on other STM32 boards?

**Comment by gschorcht** [comment]
> I'd suggest just to call `pm_block(PM_MODE_EM2)` once in `_usbdev_init()` until the clock source switching is implemented.
> 
> Once the clock source switching is implemented, we need the following logic:
> 
> * `_usbdev_init()` blocks `PM_MODE_EM2` -> EM1 is the lowest allowed pm mode
> * `_sleep_periph()` unblocks `PM_MODE_EM2` and blocks `PM_MODE_EM3` -> EM2 ist the lowest allowed pm mode
> * `_wake_periph()` unblocks `PM_MODE_EM3` and blocks `PM_MODE_EM2` -> EM1 ist the lowest allowed pm mode

@jue89 I have rebased the PR after your PR was merged and implemented the clock switch.

**Comment by dylad** [comment]
> when accessing *reset_addr due to a hard fault. Has someone observed something like that on other STM32 boards?

I am not really familiar with STM32 and I've never experiment this but since `RIOTBOOT_MAGIC_NUMBER` is supposed to be placed at the very end of the SRAM and IIRC, there are several SRAM bank in some ST families, I think this should be a starting point for debugging.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c [hunk 6]
```diff
@@ -578,15 +580,41 @@ static void _sleep_periph(const dwc2_usb_otg_fshs_config_t *conf)
...
+    CMU->USBCTRL &= ~_CMU_USBCTRL_USBCLKSEL_MASK;
+    CMU->USBCTRL |= (CLOCK_LFA == cmuSelect_LFXO) ? CMU_USBCTRL_USBCLKSEL_LFXO
+                                                  : CMU_USBCTRL_USBCLKSEL_LFRCO;
+#elif defined(CPU_FAM_EFM32GG) || defined(CPU_FAM_EFM32LG)
+    CMU->CMD &= ~_CMU_CMD_USBCCLKSEL_MASK;
+    CMU->CMD |= (CLOCK_LFA == cmuSelect_LFXO) ? CMU_CMD_USBCCLKSEL_LFXO
+                                              : CMU_CMD_USBCCLKSEL_LFRCO;
+#else
+#error "EFM32 family not yet supported"
+#endif
```
Comment (jue89) [inline drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:597]: The emlib provided by the `gecko_sdk` has an helper for clock source selection. Without having this tested, this might work for your CPU and probably will work for future CPUs, too.

```suggestion
    CMU_ClockSelectSet(cmuClock_USB, CLOCK_LFA);
```

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c [hunk 7]
```diff
@@ -578,15 +580,41 @@ static void _sleep_periph(const dwc2_usb_otg_fshs_config_t *conf)
...
+    /* switch USB core clock source either to LFXO or LFRCO */
+#if defined(CPU_FAM_EFM32GG12B)
+    CMU->USBCTRL &= ~_CMU_USBCTRL_USBCLKSEL_MASK;
+    CMU->USBCTRL |= CMU_USBCTRL_USBCLKSEL_USHFRCO;
+#elif defined(CPU_FAM_EFM32GG) || defined(CPU_FAM_EFM32LG)
+    CMU->CMD &= ~_CMU_CMD_USBCCLKSEL_MASK;
+    CMU->CMD |= CMU_CMD_USBCCLKSEL_HFCLKNODIV;
+#else
+#error "EFM32 family not yet supported"
+#endif
```
Comment (jue89) [inline drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:617]: Maybe, `CMU_ClockSelectSet()` could be used here, too.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c [hunk 8]
```diff
@@ -674,6 +705,41 @@ static void _usbdev_init(usbdev_t *dev)
...
+#elif defined(CPU_FAM_EFM32GG) || defined(CPU_FAM_EFM32LG)
+    /* enable USB system clock */
+    CMU->HFCORECLKEN0 |= CMU_HFCORECLKEN0_USB;
+    /* select HFCLK as USB PHY clock source */
+    CMU->CMD |= CMU_CMD_USBCCLKSEL_HFCLKNODIV;
+    /* enable USB core clock */
+    CMU->HFCORECLKEN0 |= CMU_HFCORECLKEN0_USBC;
+#else
+#error "EFM32 family not yet supported"
+#endif
```
Comment (jue89) [inline drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:733]: Maybe, `CMU_ClockSelectSet()` could be used here, too.

**Review by jue89 — CHANGES_REQUESTED** [review]
Regarding `pm_layered` modes: This driver is initialized in the awake state and might be put into sleep by calling `_sleep_periph()` and woken-up again by calling `_wake_periph()`, right?

File: cpu/efm32/include/periph_cpu.h [hunk 9]
```diff
@@ -42,6 +43,13 @@ extern "C" {
 #endif
 
 #if (defined(ADC_COUNT) && (ADC_COUNT > 0)) || defined(DOXYGEN)
+/**
+ * @name    Power management configuration
+ * @{
+ */
+#define PROVIDES_PM_OFF     /**< pm_off function is provided */
+/** @} */
+
```
[the remark this answers is not in the document]
Comment (gschorcht) [reply cpu/efm32/include/periph_cpu.h]: This change was added with commit 9fc917580e8683a504ee517833144771b3940770 to fix a compilation issue in CI:
`cpu/cortexm_common/include/cpu.h` had to be included in `cpu/efm32/periph_cpu.h` so that 
`PROVIDES_PM_SET_LOWEST` is defined if only `periph_cpu.h` is included as header anywhere. Otherwise `pm_set_lowest` was defined multiple times if the `pm_layered` module is not used. Additionally, `PROVIDES_PM_OFF` had to be defined. I don't remember what application produced the compilation error but we can try it again without this change.

[the remark this answers is not in the document]
Comment (gschorcht) [reply cpu/efm32/include/periph_cpu.h]: I reverted commit 9fc917580e8683a504ee517833144771b3940770, lets see what Mudock says.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 6 above)
Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:597]: After having a look into the implementation of this function, it should work. I will test it for EFM32GG12B, but I can't test it for the other EFM32 variants.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 7 above)
Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:617]: Here we unfortunately need a different handling of the EFM32 families, because they use different HF clock sources for the USB peripherals.

```c
#if defined(CPU_FAM_EFM32GG12B)
    CMU_ClockSelectSet(cmuClock_USB, cmuSelect_USHFRCO);
#elif defined(CPU_FAM_EFM32GG) || defined(CPU_FAM_EFM32LG)
    CMU_ClockSelectSet(cmuClock_USB, cmuSelect_HFCLK);
#else
#error "EFM32 family not yet supported"
#endif
```

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 8 above)
Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:733]: Hm, I already thought about that. The code here is from the reference manuals and `CMU_ClockSelectSet` doesn't cover all required settings. It enables the oscillator if necessary and sets `CMU_USBCTRL_USBCLKSEL_USHFRCO` respectively `CMU_CMD_USBCCLKSEL_HFCLKNODIV` in the according register.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 7 above)
Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:617]: :+1: Ok, I have tested it. With these changes suspend and wake up works perfectly. Without these changes, it didn't work correclty.

File: cpu/efm32/include/periph_cpu.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (jue89) [reply cpu/efm32/include/periph_cpu.h]: It's failing ... I'd say it can be fixed by changing the `#define PROVIDES_PM_LAYERED_OFF` to `#define PROVIDES_PM_OFF` in this file. It should be the valid description how `periph_pm` is implemented ...

[the remark this answers is not in the document]
Comment (gschorcht) [reply cpu/efm32/include/periph_cpu.h]: Now I remember, the problem was the `riotbuild_dfu` test app. Probably, it is not using `pm_layered`.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 8 above)
Comment (jue89) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c]: Unfortunately, I don't have a board based on these CPUs at hand.

Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c]: Hm, who could have such a board at hand?

Comment (jue89) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c]: Maybe @basilfx has? He introduced this families into RIOT.

Comment (gschorcht) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c]: Hm, @basilfx didn't react :worried: I hope @chrysn will have some time to test it, it seems that he has a `stk3700` board.

File: cpu/efm32/include/periph_cpu.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (gschorcht) [reply cpu/efm32/include/periph_cpu.h]: > I reverted commit [9fc9175](https://github.com/RIOT-OS/RIOT/commit/9fc917580e8683a504ee517833144771b3940770), lets see what Mudock says.

Hm, after the next round of CI compilation I had to add `#include "cpu.h"` in `cpu/efm32/include/periph_cpu.h` again to have `PROVIDES_PM_SET_LOWEST` defined, otherwise `pm_set` is defined multiple times. Also, it seems that both `PROVIDES_PM_LAYERED_OFF` and `PROVIDES_PM_OFF` must be defined in `cpu/efm32/include/periph_cpu.h`. `bootloaders/riotboot_dfu` does not use `pm_layered` and requires `PROVIDES_PM_OFF`.

File: drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c (hunk 8 above)
Comment (jue89) [reply drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c]: Code-wise the efm32-part looks good to me!

**Comment by benpicco** [comment]
Please squash this one

**Comment by gschorcht** [comment]
Squashed and compiled successfully.

**Comment by gschorcht** [comment]
> Not much time for reviewing, but testing I can fit in, on stk3700 and a custom board.

@chrysn Is there a chance you could test it on your `stk3700` board?

**Comment by chrysn** [comment]
Tested with stk3700:

```
$ make -C tests/usbus_hid BOARD=stk3700 PROGRAMMER=openocd all flash term
2022-12-19 09:50:36,417 # main(): This is RIOT! (Version: 2023.01-devel-470-gc6bae-drivers/usbdev_synopsys_dwc2_efm32)
2022-12-19 09:50:39,166 # RIOT USB HID echo test
2022-12-19 09:50:39,169 # Write input to the hidraw device under /dev/hidrawX, and see if it gets echoed here

$ echo foo | sudo tee /dev/hidraw0
foo

2022-12-19 09:54:37,783 # USB_HID rx_cb: Test argument 
2022-12-19 09:54:37,784 # Msg received via USB HID: 
2022-12-19 09:54:37,785 # foo
2022-12-19 09:54:37,785 # 
```

Confirming that this works on stk3700. Thanks a lot!

**Comment by gschorcht** [comment]
> Tested with stk3700

@chrysn Many thanks for testing.

**Review by benpicco — APPROVED** [review]

**Comment by gschorcht** [comment]
What's missing to get all lights green? "bors merge" or "bors try" or anything else?

**Comment by chrysn** [comment]
bors merge

**Comment by chrysn** [comment]
Ah, and @jue89 has not cleared his review position. (Weird, I'd like to click "re-request review" but don't find the button).

Juergen, are you happy with the responses to your earlier review? Please update it.

**Review by jue89 — APPROVED** [review]
ACK from my side. Everything has been addressed.

**Comment by benpicco** [comment]
bors merge

**Comment by gschorcht** [comment]
Thanks for reviewing and testing.
